### PR TITLE
Support CONTROLLED as a circuit modifier.

### LINKED ISSUE #539: Support CONTROLLED as a circuit modifier
Only DAGGER is currently allowed.

### PR DESCRIPTION
**Pull request by jmbr**
Closes #539.

### CODE REVIEW

File: src/ast.lisp [hunk 1]
```diff
@@ -1151,6 +1151,15 @@ consists of a DAGGER-OPERATOR acting on a NAMED-OPERATOR."
     ((dagger-operator dod) (plain-operator-p dod))
     (_ nil)))
 
+(defun simple-controlled-operator-p (od)
+  "Is the operator description OD a simple controlled application?
+
+An operator is considered to be a simple controlled application if it
+consists of a CONTROLLED-OPERATOR acting on a NAMED-OPERATOR."
+  (adt:match operator-description od
+    ((controlled-operator dod) (plain-operator-p dod))
```
Comment (appleby) [inline src/ast.lisp]: `DOD` -> `COD` ?  or maybe `INNER-MATRYOSHKA` (and/or `BABUSHKA`)

Comment (jmbr) [reply src/ast.lisp]: Good point!

File: src/analysis/expansion.lisp [hunk 2]
```diff
@@ -191,6 +191,43 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
            ;; The Hermitian transpose reverses the order of operator
            ;; applications
            (setf instrs (nreverse instrs))))
+        ((simple-controlled-operator-p (application-operator instr))
+         (let ((arguments (arguments instr)))
+           (when (zerop (length arguments))
+             (error "CONTROLLED requires at least one select qubit in ~/quil:instruction-fmt/." instr))
+
+           (destructuring-bind (select-qubit . data-qubits) arguments
+             ;; We call these select and data qubits following the naming
```
Comment (notmgsk) [inline src/analysis/expansion.lisp]: The rest of `cl-quil` calls them "control" and "target"

File: src/analysis/expansion.lisp [hunk 3]
```diff
@@ -191,6 +191,43 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
...
+             ;; scheme of Shende et al.
+             (when (member select-qubit data-qubits :test #'equalp)
+               (error "CONTROLLED cannot be applied when the select qubit is among the data qubits ~
+                       in ~/quil:instruction-fmt/" instr))
+
+             (let ((definition (circuit-application-definition instr)))
+               ;; Check that the instructions in the body of a given circuit
+               ;; do not involve the select qubit.
+               (dolist (x (circuit-definition-body definition) (values))
```
Comment (notmgsk) [inline src/analysis/expansion.lisp]: I'm not sure this is necessary. The same check will be performed at circuit-expansion time.

Comment (notmgsk) [reply src/analysis/expansion.lisp]: The above might be untrue actually. 🤷‍♂

File: src/analysis/expansion.lisp (hunk 2 above)
Comment (stylewarning) [reply src/analysis/expansion.lisp]: we should call the qubits `juan` and `company`

File: src/analysis/expansion.lisp [hunk 4]
```diff
@@ -191,6 +191,43 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
...
            (setf instrs (nreverse instrs))))
+        ((simple-controlled-operator-p (application-operator instr))
+         (let ((arguments (arguments instr)))
+           (when (zerop (length arguments))
+             (error "CONTROLLED requires at least one select qubit in ~/quil:instruction-fmt/." instr))
+
+           (destructuring-bind (select-qubit . data-qubits) arguments
+             ;; We call these select and data qubits following the naming
+             ;; scheme of Shende et al.
+             (when (member select-qubit data-qubits :test #'equalp)
```
Comment (stylewarning) [inline src/analysis/expansion.lisp]: there is a `qubit=`

File: src/analysis/expansion.lisp [hunk 5]
```diff
@@ -191,6 +191,43 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
...
+               (error "CONTROLLED cannot be applied when the select qubit is among the data qubits ~
+                       in ~/quil:instruction-fmt/" instr))
+
+             (let ((definition (circuit-application-definition instr)))
+               ;; Check that the instructions in the body of a given circuit
+               ;; do not involve the select qubit.
+               (dolist (x (circuit-definition-body definition) (values))
+                 (when (member select-qubit (arguments x) :test #'equalp)
```
Comment (stylewarning) [inline src/analysis/expansion.lisp]: `qubit=`

File: src/analysis/expansion.lisp [hunk 6]
```diff
@@ -191,6 +191,43 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
...
+               ;;         I ⊕ U = (I ⊕ Uₖ) ... (I ⊕ U₂) (I ⊕ U₁).
+               ;; If the select qubit is not the most significant one, we can
+               ;; reduce this to the previous case by applying suitable
+               ;; permutation matrices.
+               (let ((instrs (instantiate-instruction definition params data-qubits)))
+                 (dolist (instr instrs instrs)
+                   (unless (unitary-instruction-p instr)
+                     (error "CONTROLLED cannot be applied to the impure instruction ~
+                            ~/quil:instruction-fmt/" instr))
+                   (setf (application-operator instr) (controlled-operator (application-operator instr))
```
Comment (stylewarning) [inline src/analysis/expansion.lisp]: Living dangerously! :)

File: src/analysis/expansion.lisp (hunk 3 above)
Comment (jmbr) [reply src/analysis/expansion.lisp]: Unfortunately, if I disable these checks then the circuit is silently expanded despite the conflict.

Comment (notmgsk) [reply src/analysis/expansion.lisp]: I should have trusted you from the start :)

Comment (jmbr) [reply src/analysis/expansion.lisp]: Nah, skepticism is healthy. Even more so in code reviews :-)

**Comment by jmbr** [comment]
According to the Quil specification, multiple DAGGER and CONTROLLED modifiers should be allowed. Our current code does not let users do that, so I've filed #545 for this.

File: src/analysis/expansion.lisp (hunk 4 above)
Comment (jmbr) [reply src/analysis/expansion.lisp]: Thanks. Went with `argument=` to deal with formal parameters as well in the end.

File: src/analysis/expansion.lisp [hunk 7]
```diff
@@ -191,6 +191,41 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
            ;; The Hermitian transpose reverses the order of operator
            ;; applications
            (setf instrs (nreverse instrs))))
+        ((simple-controlled-operator-p (application-operator instr))
+         (let ((arguments (arguments instr)))
+           (when (zerop (length arguments))
```
Comment (stylewarning) [inline src/analysis/expansion.lisp]: `endp` but i'll let it slide

File: src/analysis/expansion.lisp [hunk 8]
```diff
@@ -191,6 +191,41 @@ An instruction is unitary if it is of type APPLICATION, whether that be INSTR it
...
+
+           (destructuring-bind (control-qubit . target-qubits) arguments
+             (when (member control-qubit target-qubits :test #'argument=)
+               (error "CONTROLLED cannot be applied when the control qubit is among the target qubits ~
+                       in ~/quil:instruction-fmt/" instr))
+
+             (let ((definition (circuit-application-definition instr)))
+               ;; Check that the instructions in the body of a given circuit
+               ;; do not involve the control qubit.
+               (dolist (x (circuit-definition-body definition) (values))
```
Comment (stylewarning) [inline src/analysis/expansion.lisp]:
```suggestion
               (dolist (x (circuit-definition-body definition))
```

i consider this line noise

File: tests/translator-tests.lisp [hunk 9]
```diff
@@ -316,7 +316,7 @@ U(time) 2 1 0"))
                                               (generate-translator-test-case-for-simple-compiler
                                                (quil::compiler-bindings compiler))))
                  (test-case-too-complicated (c)
-                   (format t "~&Compiler ~A not tested because ~A" compiler (test-case-too-complicated-reason c))
```
Comment (stylewarning) [inline tests/translator-tests.lisp]: this is in master and shouldn't be here

Comment (jmbr) [reply tests/translator-tests.lisp]: Drat.

**Pull request merged**